crates/sui-core/src/stake_aggregator.rs

```diff
@@ -115,21 +120,62 @@ impl<const STRENGTH: bool> StakeAggregator<AuthoritySignInfo, STRENGTH> {
                     self.data.values().cloned().collect(),
                     self.committee(),
                 ) {
-                    Ok(aggregated) => InsertResult::QuorumReached(aggregated),
+                    Ok(aggregated) => {
+                        match aggregated.verify_secure(
+                            &data,
+                            Intent::default().with_scope(T::SCOPE),
+                            self.committee(),
+                        ) {
+                            // In the happy path, the aggregated signature verifies ok and no need to verify
+                            // individual.
+                            Ok(_) => InsertResult::QuorumReached(aggregated),
+                            Err(_) => {
+                                // If the aggregated signature fails to verify, fallback to iterating through
+                                // all signatures and verify individually. Decrement total votes and continue
+                                // to find new authority for signature to reach the quorum.
+                                //
+                                // TODO(joyqvq): It is possible for the aggregated signature to fail every time
+                                // when the latest one single signature fails to verify repeatedly, and trigger
+                                // this for loop to run. This can be optimized by caching single sig verification
+                                // result only verify the net new ones.
+                                let mut bad_votes = 0;
+                                for (name, sig) in &self.data.clone() {
+                                    if let Err(err) = sig.verify_secure(
+                                        &data,
+                                        Intent::default().with_scope(T::SCOPE),
+                                        self.committee(),
+                                    ) {
+                                        // TODO(joyqvq): Currently, the aggregator cannot do much with an authority
```

Comment:
Could we just record faulty validators and not accept a new signature from it, at least for this transaction?

Reply:
its considered final as nonretryable stake, so the bad validator is not retried again for this tx

Reply:
It is possible that even though the stake for that validator is determined as nonretryable that there is still enough retryable stake outside of that and the tx could be determined as retryable.